Add checks to prevent value objects whose toString() method returns null from being passed to Lucene as an index value

amorgner opened the pull request:
It also catches the NPE Lucene throws in this cases and treats the value object as if it was a null value itself so it will be indexed as a null value.
Also, databases which crashed by the issue described in https://github.com/neo4j/neo4j/issues/6685 will recover successfully again.

amorgner committed:
This commit adds some null checks to prevent value objects whose toString() method returns null from being passed to Lucene as an index value.
It also catches the NPE Lucene throws in this cases and treats the value object as if it was a null value itself so it will be indexed as a null value.
Also, databases which crashed by the issue described in https://github.com/neo4j/neo4j/issues/6685 will recover successfully again.

tinwelint commented on community/lucene-index/src/main/java/org/neo4j/index/impl/lucene/IndexType.java:
```diff
@@ -67,7 +67,7 @@ public void addToDocument( Document document, String key, Object value )
         {
             Set<String> values = null;
-            if ( value != null )
+            if ( value != null && value.toString() != null )
```
It would be great to avoid calling `toString()` using something like:

```
String stringValue;
if ( value != null && (stringValue = value.toString()) != null )
{
   ... use the stringValue here instead of calling toString() again
}
```

And likewise for all the other placed. WDYT?

amorgner replied:
Yes, definitely. I'll modify the PR.

amorgner committed:
Added test for recovering with legacy indexes containing value objects whose toString() method returns null.

amorgner committed:
Removed some unnecessary null checks to simplify the fix.

amorgner commented:
I've removed the null checks in `InputType.java` again - they don't matter in the remove methods. I also added a recovery test which fails without this fix and did a little code clean up.

amorgner committed:
Added instanceof check as suggested by @tinwelint

tinwelint commented:
@amorgner Checking out your branch and playing around I'm realizing this is too late in the process to detect invalid values. Also I'm not a fan of the NPE caught in FullTxData, that could mask many other errors and just swallow.

What I'm thinking now is that this check should be in `LuceneIndex` and there only, where proper errors are thrown instead of accepting/swallowing in different places. Failing fast and clearly is almost always better. What I have in mind is something like:

```
diff --git a/community/lucene-index/src/main/java/org/neo4j/index/impl/lucene/LuceneIndex.java b/community/lucene-index/src/main/java/org/neo4j/index/impl/lucene/LuceneIndex.java
index 4729107..a32140a 100644
--- a/community/lucene-index/src/main/java/org/neo4j/index/impl/lucene/LuceneIndex.java
+++ b/community/lucene-index/src/main/java/org/neo4j/index/impl/lucene/LuceneIndex.java
@@ -112,11 +112,11 @@ public abstract class LuceneIndex implements LegacyIndex

     protected Object getCorrectValue( Object value )
     {
-        if ( value instanceof ValueContext )
-        {
-            return ((ValueContext) value).getCorrectValue();
-        }
-        return value.toString();
+        Object result = value instanceof ValueContext
+                ? ((ValueContext) value).getCorrectValue()
+                : value.toString();
+        assertValidValue( value );
+        return result;
     }

     private static void assertValidKey( String key )
@@ -127,6 +127,18 @@ public abstract class LuceneIndex implements LegacyIndex
         }
     }

+    private void assertValidValue( Object singleValue )
+    {
+        if ( singleValue == null )
+        {
+            throw new IllegalArgumentException( "Null value" );
+        }
+        if ( !(singleValue instanceof Number) && singleValue.toString() == null )
+        {
+            throw new IllegalArgumentException( "Value of type " + singleValue.getClass() + " has null toString" );
+        }
+    }
+
     /**
      * See {@link Index#remove(PropertyContainer, String, Object)} for more
      * generic documentation.
diff --git a/community/lucene-index/src/test/java/org/neo4j/index/impl/lucene/RecoveryTest.java b/community/lucene-index/src/test/java/org/neo4j/index/impl/lucene/RecoveryTest.java
index fc119b1..6507e8f 100644
--- a/community/lucene-index/src/test/java/org/neo4j/index/impl/lucene/RecoveryTest.java
+++ b/community/lucene-index/src/test/java/org/neo4j/index/impl/lucene/RecoveryTest.java
@@ -28,10 +28,12 @@ import java.io.IOException;
 import org.neo4j.graphdb.DynamicRelationshipType;
 import org.neo4j.graphdb.GraphDatabaseService;
 import org.neo4j.graphdb.Node;
+import org.neo4j.graphdb.PropertyContainer;
 import org.neo4j.graphdb.Relationship;
 import org.neo4j.graphdb.Transaction;
 import org.neo4j.graphdb.factory.GraphDatabaseSettings;
 import org.neo4j.graphdb.index.Index;
+import org.neo4j.graphdb.index.RelationshipIndex;
 import org.neo4j.helpers.collection.MapUtil;
 import org.neo4j.io.fs.FileSystemAbstraction;
 import org.neo4j.kernel.DefaultFileSystemAbstraction;
@@ -45,6 +47,7 @@ import org.neo4j.test.TestGraphDatabaseFactory;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.fail;

 import static org.neo4j.graphdb.DynamicRelationshipType.withName;

@@ -90,6 +93,29 @@ public class RecoveryTest
     }

     @Test
+    public void shouldNotAcceptValuesWithNullToString() throws Exception
+    {
+        try ( Transaction tx = db.beginTx() )
+        {
+            Node node = db.createNode();
+            Node otherNode = db.createNode();
+            Relationship rel = node.createRelationshipTo( otherNode, withName( "recovery" ) );
+
+            Index<Node> nodeIndex = db.index().forNodes( "node-index" );
+            RelationshipIndex relationshipIndex = db.index().forRelationships( "rel-index" );
+
+            // Add
+            assertAddFailsWithIllegalArgument( nodeIndex, node, "key1", new ClassWithToStringAlwaysNull() );
+            assertAddFailsWithIllegalArgument( relationshipIndex, rel, "key1", new ClassWithToStringAlwaysNull() );
+
+            // Remove
+            assertRemoveFailsWithIllegalArgument( nodeIndex, node, "key1", new ClassWithToStringAlwaysNull() );
+            assertRemoveFailsWithIllegalArgument( relationshipIndex, rel, "key1", new ClassWithToStringAlwaysNull() );
+            tx.success();
+        }
+    }
+
+    @Test
     public void testAsLittleAsPossibleRecoveryScenario() throws Exception
     {
         try ( Transaction tx = db.beginTx() )
@@ -254,4 +280,42 @@ public class RecoveryTest
             System.exit( 0 );
         }
     }
+
+    static class ClassWithToStringAlwaysNull
+    {
+        @Override
+        public String toString()
+        {
+            return null;
+        }
+    }
+
+    private <ENTITY extends PropertyContainer> void assertAddFailsWithIllegalArgument( Index<ENTITY> index,
+            ENTITY entity, String key, Object value )
+    {
+        try
+        {
+            index.add( entity, key, value );
+            fail( "Should not accept value with null toString" );
+        }
+        catch ( IllegalArgumentException e )
+        {
+            // Good
+        }
+    }
+
+    private <ENTITY extends PropertyContainer> void assertRemoveFailsWithIllegalArgument( Index<ENTITY> index,
+            ENTITY entity, String key, Object value )
+    {
+        try
+        {
+            index.remove( entity, key, value );
+            fail( "Should not accept value with null toString" );
+        }
+        catch ( IllegalArgumentException e )
+        {
+            // Good
+        }
+    }
+
 }
```

and that would be the only change.

amorgner commented:
@tinwelint That's even better, for sure. I'll test the patch against my test cases. If it works, should I apply the patch to the PR, or just withdraw the PR and let you make the changes?

tinwelint commented:
@amorgner please do, no I don't want to take credit for your effort here. Please submit a PR with these changes if they work alright for you!

amorgner committed:
Modified PR following @tinwelint's patch so that indexing an invalid value with null toString() method will fail early with an exception.

amorgner commented:
@tinwelint Finally found time to check the patch you proposed. It prevents the object with null toString from being indexed, and thus the tx log and the db will be clean, so I'm happy with the fix.

A database which is already affected by the bug won't recover as I removed the exception handling in the log applier code, but I think we can handle those individual cases.

tinwelint commented:
@amorgner oh OK, good point. So what would be the absolutely minimal addition that we can do to have existing affected transaction logs skip these?

amorgner commented:
@tinwelint Problem seems to be that at recovery, the `valueType` is `IndexCommand.VALUE_TYPE_NULL` so the index value `null` is passed to `addDocument` in `LuceneCommandApplier#visitIndexAddNodeCommand` which leads to the NPE. Is there a situation when adding `null` to an index makes any sense? On write, the null type is just ignored, so wouldn't it make sense to ignore it on read?

tinwelint commented:
@amorgner the null type participates in `Index#remove` calls where value is left out (from method signature even), which is not the same thing as ignoring values passed in which happens to be null or return a null `toString()`, so they aren't ignored as such. Although that might be a good idea to add that check, perhaps as low as in `LuceneCommandApplier` to automatically "repair" databases with this issue in its transaction log. That check's only purpose would be just that. So both in `visitIndexAddNodeCommand` and `visitIndexAddRelationshipCommand` with a good comment as to why value might be null at this point. Is that something you had in mind?

tinwelint commented:
@amorgner OK, I'll take it upon me to summarize this work and discussion in a new PR. Coming soon.

amorgner commented:
@tinwelint Sorry for not following up directly. I had already working code but had too little time to fiinalize it and then was on vacation. Should I add to this PR? I'm also fine with a new PR.

tinwelint commented:
It would be great with a shaped up PR of one or two committs or something,
please do if you have it ready.

On Tue, Apr 12, 2016 at 1:46 PM, Axel Morgner notifications@github.com
wrote:

> @tinwelint https://github.com/tinwelint Sorry for not following up
> directly. I had already working code but had too little time to fiinalize
> it and then was on vacation. Should I add to this PR? I'm also fine with a
> new PR.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/neo4j/neo4j/pull/6691#issuecomment-208860125

## 

Mattias Persson
Neo4j Hacker at Neo Technology

tinwelint commented:
Cleaned up and issued https://github.com/neo4j/neo4j/pull/7097 instead, does that look OK to you @amorgner

amorgner commented:
Thanks @tinwelint for following up here, I didn't find the time. The fix looks good so far (and much better than what I had). I'll test it with my setup and report my findings in #7097.
